Make CONFIG_USERSPACE=y and CONFIG_CMAKE_LINKER_GENERATOR=y work together

bjorniuppsala opened the pull request:
This makes the CMake linker generator work with CONFIG_USERSPACE=y

It uses the features added in #84635 to make e.g. controlling which passes include what easier (the same changes are incorporated in this PR too). 

I have added ways to funnel inputs from e.g. gen_kobject_placeholders.py and gen_app_partitions.py into the linker generator, in the ld-skeletons this is handled by using the the c preprocessor. I have tried to create general and reusable ways of gathering information from the intermediate linking passes easy and natural. Look at the "Functions for input from prev linker pass" commit for details.

I have split cmake/linker_script/arm/linker.cmake into multiple files to reflect how the ld-skeletons are structured.
There are few zephyr_linker_*() calls that have been changed and added to mirror what the ld-skeletons are doing for userspace. Despite this there are a fair number of differences and missing sections for linker generator, that are not related to userspace that remains to be fixed until it is a complete replacement for the skeletons.

This is currently only tested on gcc for cortex_m4, although I think the changes to make other cortex cores and toolchains working should not be too hard. The aim is to provide IAR toolchain support for CONFIG_USERSPACE, once the actual iar toolchain support is in.

Edit:
Note that the changes I propose here does not quite create a fully working solution. There are a number of further changes needed.
Most prominent may be proper handling of MPU_ALIGN(region_size) for sections defined by the cmake linker generator (see #86563), as is you have to use large enough values for e.g. CONFIG_CUSTOM_SECTION_MIN_ALIGN_SIZE or CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE to get proper alignment).

In the end this effort aims to provide CONFIG_USERSPACE support for the IAR toolchain

bjorniuppsala committed:
cmake: linker generator: ld: Implement MIN/MAX_SIZE for sections in ld

Implement the MIN_SIZE and MAX_SIZE options for the ld linker file
generator

Signed-off-by: Björn Bergman <bjorn.bergman@iar.com>

bjorniuppsala committed:
cmake: linker generator: Add MIN/MAX_SIZE to zephyr_linker_section()

Add MIN_SIZE and MAX_SIZE options to zephyr_linker_section()
and zephyr_linker_section_configure().
This allows padding (for MIN_SIZE) and link-time checking (for MAX_SIZE)
of sections (and parts of sections).

Clarify comments for zephyr_linker_section_configure

Signed-off-by: Björn Bergman <bjorn.bergman@iar.com>

bjorniuppsala committed:
cmake: linker generator: Handle NOT PASS and multiple PASS options

Make it possible to have multiple PASS parameters to
zephyr_linker_section() and zephyr_linker_section_configure() sections
(oring them) OR to have multiple PASS NOT p options (in which case the
sections applies in neither of the passes)

Signed-off-by: Björn Bergman <bjorn.bergman@iar.com>

bjorniuppsala committed:
cmake: linker generator: Refactor generator input file creation

Move creation of the generator arguments file from each target.cmake
into a function (zephyr_linker_generate_linker_settings_file).

Signed-off-by: Björn Bergman <bjorn.bergman@iar.com>

bjorniuppsala committed:
cmake: userspace: A stab at USERSPACE + CMAKE_LINKER_GENERATOR

All in all this makes cmake linker generator able to work for at least
most of the kernel tests and samples, on cortex m4.

Make the cmake linker generator have a file-structure more similar to
the ld-skeletons.

Add or edit groups, sections and logic to make the generator reflect
what the ld-skeletons do, esp. for kobjects and APP_SMEM, nonint and
some other details are also effected.

Use the new zephyr_linker_include_generated() and
zephyr_linker_include_var() functions to to handle
${DEVICE_API_LINKER_SECTIONS_CMAKE}, the kobject-prebuilt-*.h files and
APP_SMEM partition. Essentially the output from gen_app_partitions.py,
gen_kobject_placeholders.py.

Add ALIGN_WITH_INPUT on sections being put into DATA_REGION. This makes
the init layout work for ld.

This leverages the updates in gen_app_partitions.py to generate its
output as cmake linker generator sections too, and puts them into a
group defined in linker.cmake

Setup generator variables for alignment of APP_SMEM. Note that this does
not yet handle MPU_ALIGN which depends on the size of the section...

Fix broken k_object_assignment iterable section

Signed-off-by: Björn Bergman <bjorn.bergman@iar.com>

bjorniuppsala committed:
cmake: linker generator: Functions for input from prev linker pass

Add methods for the linker-script generator to pick up information
from previus linkerpasses:
 * zephyr_linker_include_generated(FILE f.cmake [PASS X]  [KCONFIG])
   cmake-includes f.cmake in the relevant PASSes. This allows scripts
   to generate any content (e.g. sections or section configs) as if
   it was generated by the main cmake machiery. This is intended to
   cover the case of ld-script snippets generated by e.g.
   gen_app_partitions.py
   It can also parse KCconfig files to capture @CONFIG_FOO@ variables
 * @FOO[,undef:v]@ in an value or expression in the zephyr_linker_*
   inputs. This looks for a defintion of FOO (se below), and uses its
   value if found. If FOO is not defined, the whole @@ thing is left
   untouched, unless ,undef: is used, in which case v is used instead.
 * zephyr_linker_include_generated(FILE f.h [PASS x]) greps for f.h
   for #define FOO value to be accessed as @FOO@
 * zephyr_linker_include_var(VAR FOO VALUE V) allows the main
   cmake-script to set a @FOO@ variable.

Note that the #define support is VERY basic, and does not use a proper
pre-processor. It works for the current use of files generated by e.g.
gen_kobject_placeholders.py but is not a general solution.

Signed-off-by: Björn Bergman <bjorn.bergman@iar.com>

bjorniuppsala committed:
scripts: Make gen_app_partitions.py cmake linker generator savvy

Tweak gen_app_partitions.py to be able to generate cmake linker
generator output as well as ld-script fragments.

Signed-off-by: Björn Bergman <bjorn.bergman@iar.com>

pdgendt commented on cmake/modules/extensions.cmake:
```diff
@@ -5361,6 +5362,88 @@ function(zephyr_linker_symbol)
+# PASS [NOT] <pass>: Rule for which PASSES to include file.
+#                    see zephyr_linker_section PASS
+#
+function(zephyr_linker_include_generated)
```
Same function is added twice?

bjorniuppsala replied:
How embarrassing ;)

pdgendt commented on CMakeLists.txt:
```diff
@@ -71,6 +71,12 @@ set(ZEPHYR_CURRENT_LINKER_PASS 0)
 set(ZEPHYR_CURRENT_LINKER_CMD    linker_zephyr_pre${ZEPHYR_CURRENT_LINKER_PASS}.cmd)
 set(ZEPHYR_LINK_STAGE_EXECUTABLE zephyr_pre${ZEPHYR_CURRENT_LINKER_PASS})
 
+#ensure that autoconf #defines are available to the linker script generator
```
Comments need a space after the `#` character.

bjorniuppsala replied:
Done

pdgendt replied:
The comments in this file aren't adjusted.

pdgendt commented on cmake/linker_script/common/common-noinit.cmake:
```diff
@@ -0,0 +1,9 @@
+  SYMBOLS z_user_stacks_start z_user_stacks_end)
+
+  ##include <snippets-noinit.ld>
+include(${COMMON_ZEPHYR_LINKER_DIR}/kobject-priv-stacks.cmake)
```
Add missing newline at the end.

pdgendt replied:
Not resolved yet.

pdgendt commented on cmake/linker_script/common/kobject-data.cmake:
```diff
@@ -0,0 +1,70 @@
+if(CONFIG_USERSPACE)
```
Missing license and copyright comment.

pdgendt commented on cmake/linker_script/common/kobject-priv-stacks.cmake:
```diff
@@ -0,0 +1,43 @@
+#Keep in synch with include/zephyr/linker/kobject-priv-stacks.ld
+if(CONFIG_USERSPACE)
+  if(CONFIG_GEN_PRIV_STACKS)
+    #/* Padding is needed to preserve kobject addresses
+	  #* if we have reserved more space than needed.
+	  #*/
```
Weird comment style, remove the C multi-line comment and only keep `#` comments. And fix indentation.

pdgendt commented on cmake/linker_script/common/kobject-priv-stacks.cmake:
```diff
@@ -0,0 +1,43 @@
+#Keep in synch with include/zephyr/linker/kobject-priv-stacks.ld
```
License and copyright.

pdgendt commented on cmake/linker_script/common/kobject-priv-stacks.cmake:
```diff
@@ -0,0 +1,43 @@
+      )
+    endif()
+  endif()
+endif()
```
Missing newline.

bjorniuppsala replied:
Sorry to bother you so many small misstakes, they should all be cleared by now. 
Thank you for your input

pdgendt commented on cmake/linker/ld/ld_script.cmake:
```diff
@@ -250,6 +280,19 @@ function(section_to_string)
       endif()
     endif()
+    # Min and Max size may need symbols too, even if the user did not want them
+    if(DEFINED min_size OR DEFINED max_size )
```
```suggestion
    if(DEFINED min_size OR DEFINED max_size)
```

pdgendt commented on cmake/linker/xt-ld/target.cmake:
```diff
@@ -27,6 +27,11 @@ macro(configure_linker_script linker_script_gen linker_pass_define)
   set(extra_dependencies ${ARGN})
 
   if(CONFIG_CMAKE_LINKER_GENERATOR)
+    #This should probably do something like:
+    #set(cmake_linker_script_settings ${PROJECT_BINARY_DIR}/include/generated/ld_script_settings_${linker_pass_define}.cmake)
+    #zephyr_linker_generate_linker_settings_file(FILE ${cmake_linker_script_settings})
+    #And pass the file to the generator
```
Add space after `#`

pdgendt commented on cmake/linker_script/common/common-noinit.cmake:
```diff
@@ -0,0 +1,15 @@
+# SPDX-License-Identifier: Apache-2.0
+# The contents of this file is based on include/zephyr/linker/common-noinit.ld
+# Please keep in synch
```
Also in other places of the code

```suggestion
# Please keep in sync
```

pdgendt commented on cmake/linker_script/common/common-ram.cmake:
```diff
@@ -24,77 +26,94 @@ endif()
-#if(CONFIG_USERSPACE)
-#	_static_kernel_objects_end = .;
-#endif()
+if(CONFIG_USERSPACE)
+	#_static_kernel_objects_end = .;
+endif()
```
Remove unused code

pdgendt reviewed (commented):
License headers need to be accompanied with a copyright notice, I think.

bjorniuppsala commented:
> License headers need to be accompanied with a copyright notice, I think.

I was afraid of that... The thing is that the files I am adding are either based on ld-skeletons or directly copy-pasted from linker.cmake, and there are missing copyright tags on many of these sources. So it is "hard" to add a fair copyright notice.

There are also many files (e.g. among include/linker/common/*.ld) that have license tag but not copyright.

pdgendt commented on cmake/linker/linker_script_common.cmake:
```diff
@@ -617,6 +649,83 @@ endfunction()
+  string(REGEX MATCHALL "@([^@]*)@" match_res "${src}")
+  foreach(match IN LISTS match_res)
+    string(REPLACE "@" "" expr ${match})
+    #the variable expression is as follows: s
+    #@NAME[,undef:VALUE]@ Where the VALUE gets picked if we dont find NAME
```
Add space. Also is the trailing `s` on the first line intended?

bjorniuppsala replied:
Done

pdgendt commented on cmake/linker_script/arm/linker.cmake:
```diff
@@ -119,13 +146,35 @@ zephyr_linker_section_configure(SECTION .rodata ALIGN 4)
+  zephyr_linker_section_configure(SECTION .bss INPUT COMMON)
+  zephyr_linker_section_configure(SECTION .bss INPUT ".kernel_bss.*")
+
+  #ifdef CONFIG_CODE_DATA_RELOCATION
+  #include <linker_sram_bss_relocate.ld>
+  #endif
```
Remove unused code

pdgendt commented on cmake/linker_script/common/kobject-data.cmake:
```diff
@@ -0,0 +1,63 @@
+      )
+  endif()
+
+  # During LINKER_KOBJECT_PREBUILT and LINKER_ZEPHYR_PREBUILT,
+	# space needs to be reserved for the rodata that will be
+	# produced by gperf during the final stages of linking.
+	# The alignment and size are produced by
+	# scripts/build/gen_kobject_placeholders.py. These are here
+	# so the addresses to kobjects would remain the same
+	# during the final stages of linking (LINKER_ZEPHYR_FINAL).
```
Change indentation from tabs to spaces

pdgendt commented on cmake/linker_script/common/kobject-data.cmake:
```diff
@@ -0,0 +1,63 @@
+    zephyr_linker_section_configure(
+      SECTION kobject_data
+      SYMBOLS _thread_idx_map
+      #MIN_SIZE @KOBJECT_DATA_SZ@ #
```
Remove unused code

pdgendt commented on cmake/linker_script/common/kobject-priv-stacks.cmake:
```diff
@@ -0,0 +1,42 @@
+if(CONFIG_USERSPACE)
+  if(CONFIG_GEN_PRIV_STACKS)
+    # Padding is needed to preserve kobject addresses
+	  # if we have reserved more space than needed.
```
Tab tos spaces

pdgendt commented on cmake/linker_script/common/kobject-priv-stacks.cmake:
```diff
@@ -0,0 +1,42 @@
+    )
+
+    # During LINKER_KOBJECT_PREBUILT and LINKER_ZEPHYR_PREBUILT,
+	  # space needs to be reserved for the rodata that will be
+	  # produced by gperf during the final stages of linking.
+	  # The alignment and size are produced by
+	  # scripts/build/gen_kobject_placeholders.py. These are here
+	  # so the addresses to kobjects would remain the same
+	  # during the final stages of linking (LINKER_ZEPHYR_FINAL).
```
Tabs to spaces

pdgendt commented on cmake/linker_script/common/kobject-rom.cmake:
```diff
@@ -0,0 +1,21 @@
+# Please keep in sync
+
+# During LINKER_KOBJECT_PREBUILT and LINKER_ZEPHYR_PREBUILT,
+#	space needs to be reserved for the rodata that will be
+#	produced by gperf during the final stages of linking.
+#	The alignment and size are produced by
+#	scripts/build/gen_kobject_placeholders.py. These are here
+#	so the addresses to kobjects would remain the same
+#	during the final stages of linking (LINKER_ZEPHYR_FINAL).
```
Same

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -5361,6 +5361,36 @@ function(zephyr_linker_symbol)
+# Usage:
+#   zephyr_linker_generate_linker_settings_file([FILE <name>] [STRING <string>])
+#
+# Add the content generated so far by the zephyr_linker_* functions to a file
```
`generated so far` is not correct.

When using `file(GENERATE ...)` then content is not written until CMake generator time, which again means that settings AFTER `zephyr_linker_generate_linker_settings_file()` is also taken into consideration.

For example
```
zephyr_linker_generate_linker_settings_file(FILE foo.txt)
zephyr_linker_symbol(SYMBOL __foo EXPR "(@__bar@)")
```

will actually handle the late `zephyr_linker_symbol()` call also and thus allowing applications to adjust or create symbols and sections as they like.

This is the reason generator expression are used.

bjorniuppsala replied:
Well spotted! And fixed.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -5361,6 +5361,36 @@ function(zephyr_linker_symbol)
+# or string variable.
+#
+# FILE <name>  : File to be created
+# STRING <string>: String variable to be set
```
if wanting to support string output, then it should be very clear what should be expected of this string and how it may be used.

So far my feeling is that you have used this option for debug purposes, correct ?

bjorniuppsala replied:
Yes, I have removed it now.

tejlmand commented on cmake/linker/linker_script_common.cmake:
```diff
@@ -638,6 +649,83 @@ endfunction()
+      # can't warn here because we can't check for what is relevant in this pass
+      # message(WARNING "Missing definition for ${match}")
+    endif()
+    message("Replacing ${match} with ${value}")
```
I could imagine this would result in a lot of general message output which the user has no idea what is about or what to use with it.

bjorniuppsala replied:
True! I guarded this on CMAKE_VERBOSE_MAKEFILE

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -5373,6 +5427,9 @@ endfunction()
 function(zephyr_linker_generate_linker_settings_file)
   set(single_args "FILE;STRING")
   cmake_parse_arguments(GEN "" "${single_args}" "" ${ARGN})
+  get_property(list_of_variables TARGET linker PROPERTY VARIABLES)
+  list(JOIN list_of_variables "\n" VARS)
```
This breaks one of the designs of the linker generator because it now become configure time order dependent, meaning:
```
zephyr_linker_generate_linker_settings_file(FILE foo.txt)
zephyr_linker_include_var(VAR foo)
```
will not take `foo` into consideration when generating input to the actual linker script generator executed at build time.

Perhaps not an issue just now, but may be harder to handle in future if not addressed now.

```

bjorniuppsala replied:
I changed so that the variable values are handled just like the other things so that foo in your example will be there. 
I think this may cause us to want to have PASS filtering on the variables too, if nothing else than making it possible to have different values in different passes.

tejlmand commented on CMakeLists.txt:
```diff
@@ -71,6 +71,13 @@ set(ZEPHYR_CURRENT_LINKER_PASS 0)
 set(ZEPHYR_CURRENT_LINKER_CMD    linker_zephyr_pre${ZEPHYR_CURRENT_LINKER_PASS}.cmd)
 set(ZEPHYR_LINK_STAGE_EXECUTABLE zephyr_pre${ZEPHYR_CURRENT_LINKER_PASS})
 
+# Make autoconf #defines are available to the linker script generator
```
Why not load the `.config` directly instead of going through `autoconf.h` ?
`autoconf.h` is created from the content of `.config`.

The `.config` can be loaded into the linker script generator using:
```
import_kconfig(CONFIG ${CMAKE_CURRENT_BINARY_DIR}/zephyr/.config)
```

bjorniuppsala replied:
Ah! I had no idea there was such a thing. 
It would make sense to put import_kconfig() straight into linker_script_common.cmake, right ?

bjorniuppsala replied:
I really dislike the idea of picking #defines from files. I deally the infodug up by the python macinery should be imported thorugh a nicer format, (from which the header files could be generated too if needed). But I could just not bring my self to do this at the moment.

bjorniuppsala replied:
I tried to include (..../extensions.cmake) to get import_kconfig. It does not work (due to dependency "hell"; user_cache,python,CheckCCompilerFlag,....).
find_package(Zephyr ...) also does not work due to "stuff" not being defined (BOARD not defined being the first).

The most reasonable way to do this is to break out some parts of the extensions.cmake into a smaller package with fewer requirements and import that. 

I think this might be the right way to go, but it is also a fairly long detour for a quite improvement. So I suggest that we do this later (tm).

bjorniuppsala replied:
I took the effort to move the kconfig functions from extensions.cmake into its own file. and included that into linker_script_common.cmake. it works, but is it worth it ?

bjorniuppsala replied:
Earlier I tried include without the CMAKE_MODULE_PATH. Thats where i found my dependency hell.

I think the find_package way looks cleaner. Both works. Which do you prefer ? Is there any real benefit from either ?

bjorniuppsala replied:
It works, but it also requires ZEPHYR_BASE to be set up since just loading extensions.cmake brings in user_cache & python. Adding all of 300ms to the runtime ;)

bjorniuppsala replied:
Without ZEPHYR_BASE user_cache emits some noise:
```
-- Found Python3: /usr/bin/python (found suitable version "3.10.12", minimum required is "3.10") found components: Interpreter 
/usr/bin/python: can't open file '/scripts/build/dir_is_writeable.py': [Errno 2] No such file or directory
-- Cache files will be written to: /.cache
```
From around user_cache.cmake:64

One can set USER_CACHE_DIR to get around most of this.

I dont mind the extra time really so to keep on topic I'll just set the the required stuff.

bjorniuppsala replied:
Now, the latest push tries to mitigate this.

bjorniuppsala replied:
I did not mean to sound sloppy, of course we should avoid any extra build-time if can. That is one of the reasons I checked how much extra time it took to include all of extensions.cmake.

I seem to "have bad luck" when trying find_package(). It complains about not finding FindZephyr.cmake unless I have: 
`find_package(Zephyr REQUIRED COMPONENTS extensions **HINTS $ENV{ZEPHYR_BASE}**)`, despite the CMAKE_MODULE_PATH 

I guess that `west export-zephyr` fixes that, which way is to be prefered ?

bjorniuppsala replied:
With some further thinking, I started to worry that the relying on ZEPHYR_BASE or other default path machinery might interfere in bad ways if you have multiple zephyr directories. I think it is rather important that we guarantee that we only try to find stuff in the same zephyr-tree as the linker generator itself comes from, right?

tejlmand replied:
> I guess that `west export-zephyr` fixes that, which way is to be prefered ?

Oh my, someone didn't read the `getting started` :laughing: 
https://docs.zephyrproject.org/latest/develop/getting_started/index.html
![image](https://github.com/user-attachments/assets/3df9a60c-f3b6-45b4-a57c-cfca3250a98d)

> I started to worry that the relying on ZEPHYR_BASE or other default path machinery might interfere in bad ways if you have multiple zephyr directories.

`find_package(Zephyr)` will use the following search order, so it will prefer the Zephyr found in the same hierarchy as this file: https://docs.zephyrproject.org/latest/build/zephyr_cmake_package.html#zephyr-cmake-package-search-order

But as we know where we are (inside Zephyr repo), then we can use the hint of ourselves, like this:
```
find_package(Zephyr REQUIRED COMPONENTS extensions HINTS ${CMAKE_CURRENT_LIST_DIR}/../../)
```
and thereby make sure which Zephyr we use.

Personally I think both solutions are fine:
```
find_package(Zephyr REQUIRED COMPONENTS extensions HINTS ${CMAKE_CURRENT_LIST_DIR}/../../)
```
or
```
# Remember to invoke the linker generator with `-DZEPHYR_BASE=${ZEPHYR_BASE}` in addition.
list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/../modules/")
include(extensions)
```

bjorniuppsala replied:
Does this mean that this conversation is resolved ?

tejlmand reviewed (changes requested):
Thanks for this, a lot of great work.

First round of review comments ready, though I haven't been through everything in details so I might notice more things later.
I also want to carry out some testing on this.

So far mostly minor observations, where the only main concern found is the use of `get_property()` which means code becomes CMake configure time process order dependent and thereby prevents application code to customize / extend the linker script generated.

bjorniuppsala commented on cmake/modules/extensions.cmake:
```diff
@@ -5361,6 +5361,36 @@ function(zephyr_linker_symbol)
 endfunction()
 
+# Usage:
+#   zephyr_linker_generate_linker_settings_file([FILE <name>] [STRING <string>])
```
Yep, I removed the string thing.

bjorniuppsala commented on cmake/linker/xt-ld/target.cmake:
```diff
@@ -27,6 +27,11 @@ macro(configure_linker_script linker_script_gen linker_pass_define)
   set(extra_dependencies ${ARGN})
 
   if(CONFIG_CMAKE_LINKER_GENERATOR)
+    # This should probably do something like:
+    # set(cmake_linker_script_settings ${PROJECT_BINARY_DIR}/include/generated/ld_script_settings_${linker_pass_define}.cmake)
+    # zephyr_linker_generate_linker_settings_file(FILE ${cmake_linker_script_settings})
+    # And pass the file to the generator
+
```
Ill do that instead

bjorniuppsala commented on cmake/modules/extensions.cmake:
```diff
@@ -5385,6 +5376,30 @@ macro(zephyr_linker_arg_val_list list arguments)
+# Internal helper that checks if we have consistent PASS arguments.
+# If the first element is NOT, check that all even elements are NOT, otherwise
+# check that no element is NOT
+macro(zephyr_linker_check_pass_param function_name SECTION_PASS)
```
I left the function. Maybe it is too ambitious to do these checks ?

bjorniuppsala commented on cmake/linker/linker_script_common.cmake:
```diff
@@ -94,15 +94,34 @@ function(create_group)
-    if(NOT (${SECTION_PASS} IN_LIST PASS))
-      # This section is not active in this pass, ignore.
+    is_active_in_pass(active ${PASS} "${SECTION_PASS}")
+    if(NOT ${active})
```
Thanks!

bjorniuppsala committed:
cmake: linker generator: ld: Fix alignment bug

It seems as if ld_script.cmake hasnt handled
zephyr_linker_section_configure( ALIGN ...) due to a typo
in a variable name,

Signed-off-by: Björn Bergman <bjorn.bergman@iar.com>

bjorniuppsala committed:
cmake: linker generator: ld: Files for section input patterns

For CONFIG_USERSPACE the input from gen_app_partitions.py there is
a need to be able to specify input files as well as input sections
patterns for zephyr_linker_section_configure().  This is used for
app partitions from libraries (which generate input patterns like
foo.a:*(.data*)).

This adds documentation to zephyr_linker_section_configure() to clarify
what INPUT allows, and also adds some parsing tricks to ld_script.cmake
to properly add the file patterns when they are missing.

Signed-off-by: Björn Bergman <bjorn.bergman@iar.com>

tejlmand commented on cmake/linker/linker_script_common.cmake:
```diff
@@ -617,6 +651,97 @@ endfunction()
+    message("Reading file ${INC_FILE}")
+    if(INC_KCONFIG)
+      import_kconfig(CONFIG ${INC_FILE})
+    elseif(${INC_FILE} MATCHES "^.*\.cmake$")
+      include(${INC_FILE})
+    elseif(${INC_FILE} MATCHES "^.*\.h$")
+      list(APPEND PREPROCESSOR_FILES ${INC_FILE})
```
this is hiding what filetypes is actually supported, perhaps it's better to be explicit here:
https://github.com/zephyrproject-rtos/zephyr/blob/d86f7031b9377e20de99a6c7d311cdfb66a488d3/cmake/modules/extensions.cmake#L5314-L5317

So that the developers wanting to use `zephyr_linker_include_generated()` can actually see directly that Kconfig, c-headers, and cmake files are supported.

bjorniuppsala replied:
Fixed (by replacing the FILE tag with KCONFIG, HEADER or CMAKE)

tejlmand commented on CMakeLists.txt:
```diff
@@ -71,6 +71,15 @@ set(ZEPHYR_CURRENT_LINKER_PASS 0)
+# Make kconfig variables available to the linker script generator
+zephyr_linker_include_generated(FILE ${CMAKE_CURRENT_BINARY_DIR}/.config KCONFIG)
+
+# The linker generator also needs sections.h to be able to access e.g. _APP_SMEM_SECTION_NAME
+# for linkerscripts that do not support c-preprocessing.
+zephyr_linker_include_generated(FILE ${ZEPHYR_BASE}/include/zephyr/linker/sections.h)
```
not to be changed in this PR.

I consider that perhaps it's better to have developers being able to describe the content of `sections.h` in CMake or a template file which CMake can use, and from there generate header files, CMake files, and whatever else is needed by the build system.

CMake is better at generating header files than reading them :wink: 

Thoughts ?

bjorniuppsala commented on cmake/modules/kconfig_extensions.cmake:
I will, i will. Take it easy ;)

bjorniuppsala commented:
Going from the current state of this PR there is a "big"question left, ohw to handle MPU_ALIGN(region_size). 

I deffered that to an RFC: #86563, hoping for some more attention from interested parties.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -5393,6 +5384,96 @@ function(zephyr_linker_symbol)
+    message(FATAL_ERROR "zephyr_linker_include_generated(${ARGV0} ...) must have one of CMAKE, KCONFIG or HEADER")
+  endif()
+
+  if(DEFINED INCLUDE_PASS)
+    zephyr_linker_check_pass_param("${INCLUDE_PASS}")
+  endif()
```
afaict what you want to check is that `NOT` is only allowed as the first entry and not as subsequent values, like this:
`INCLUDE_PASS=NOT;SOME;PASS;VALUES`
but you don't care if there is a `NOT` or not.
So those are fine as well:
`INCLUDE_PASS=SOME;PASS;VALUES`
`INCLUDE_PASS=` (see other comment regarding empty values to an argument)
but those should be caught:
`INCLUDE_PASS=SOME;PASS;NOT;VALUES`
`INCLUDE_PASS=NOT;SOME;PASS;NOT;VALUES`

which you can do much easier like this:
```suggestion
  list(SUBLIST INCLUDE_PASS 1 -1 check)
  if("NOT" IN_LIST check)
    message(FATAL_ERROR "NOT only allowed before first <pass> value, like this: PASS [NOT] <pass>...")
  endif()
```

Note: We don't care if `INCLUDE_PASS` is defined or not, because if `INCLUDE_PASS` is undefined or empty, then `check` will also be an empty list when using `SUBLIST`, and thus not contain `NOT`.
Also we don't care the value of the first element, what we care about is that no element after first element is `NOT`.

bjorniuppsala replied:
I kept the function, but simplified it using your suggestions.

tejlmand reviewed (commented):
Thanks for the update. Looks good and close to +1 .

Some minor observations, mainly related to the `zephyr_linker_check_pass_param()`.
A function which we may even be able to replace with a very simple `list()` commadn and `if()` check.
Feel free to reply to observations.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -5413,6 +5494,24 @@ macro(zephyr_linker_arg_val_list list arguments)
+    # The first NOT is ok, drop it for the coming checks
+    list(SUBLIST SECTION_PASS 1 -1 SECTION_PASS)
+  endif()
+  if(NOT SECTION_PASS)
+    message(FATAL_ERROR "At least one pass name required")
```
Those are fair points, and if checking for values to `PASS` argument is a conscious choice, then I approve the decision.
\
But please make this restriction clear in the description of the function.

bjorniuppsala replied:
In the end I think we'll stick with your proposal. PASS <empty list> means not active anywhere. This should be "safe" from a waste perspective...

tejlmand reviewed (dismissed):
LGTM, no blocking comments.

But please take a look at this comment: https://github.com/zephyrproject-rtos/zephyr/pull/85241#discussion_r1993781928

bjorniuppsala commented:
I would appreciate input or approvals.

RobinKastberg reviewed (approved):
LGTM

bjorniuppsala commented:
With CONFIG_CMAKE_LINKER_GENERATOR=y  test/kernel/threads/tls is still broken due to missing alignment of _t{data,bss}_size. 
There is currently no way to express the end-alignment that works for ld.  

Given that we lack MPU_ALIGN anyways, I'll work on a followup as a separate PR.
